src/mappings/remark.ts

```diff
@@ -3,49 +3,113 @@ import { CallHandlerContext } from '../processor'
 import events from '../types/events'
 import { config } from '../config'
 import { TokenTransfer, Transfer } from '../model'
+import { Call } from '@subsquid/substrate-processor'
 
-export async function handleSystemRemark(ctx: CallHandlerContext) {
+export async function handleBatchWithRemark(ctx: CallHandlerContext) {
     let origin = codec(config.prefix).encode(ctx.call.origin.value.value)
 
-    let matchingBalanceTransfer = await ctx.store.findOne(Transfer, {
-        where: {
-            from: origin,
-            blockNumber: ctx.call.block.height,
-        },
-    })
-
-    let matchingTokenTransfer = await ctx.store.findOne(TokenTransfer, {
-        where: {
-            from: origin,
-            blockNumber: ctx.call.block.height,
-        },
-    })
-
-    // Error if found more than one on the union of both token and balances, or none
-    if (matchingBalanceTransfer && matchingTokenTransfer) {
-        ctx.log.info(
-            `Error when finding the correct transfer for remark with id ${ctx.call.id}. Inconsistent number of transfers for user: ${origin}`
-        )
+    // filter out batch calls where there one transfer and
+    // one remark only
+    if (!hasSystemWithRemark(ctx.call)) {
         return
     }
+    const calls = ctx.call.args.calls
+    const kinds = calls.map((call: any) => call.__kind)
+    if (kinds.length != 2) return
+
+    const remarkCall = calls.find((call: any) => call.__kind === 'System')
 
-    let remarkRaw = ctx.call.args.remark
+    let remarkRaw = remarkCall.value.remark
     let remark = null
+    // Try decoding remark from hex
     try {
-        // Try decoding remark from hex
         remark = Buffer.from(remarkRaw.slice(2), 'hex').toString('utf8')
     } catch (e) {
         ctx.log.info(
             `Error when decoding remark '${remarkRaw}' with id ${ctx.call.id}.`
         )
     }
 
-    // Either we append the remark to balance transfer or token
-    if (matchingBalanceTransfer) {
+    // handle balances pallet transfer with remark
+    if (kinds.includes('Balances')) {
```

Comment:
We have a lot of duplicated code in both if-else branches, let's try to deduplicate it with some refactoring.